### ISSUE TITLE
Compute training loss in NMFTrainer separately from gradient computation

### ISSUE THREAD

**Issue by yunseong** [snuspl/cruise#945]
In NMFTrainer, the sum of the training loss is computed on-line, while computing gradient. This is possible because the training data each Trainer processes only affects the different part of models, independently of other Trainers. In other apps (e.g., MLR, LDA), on the other hand, the entire model is pulled after all training data are processed, to reflect the changes from other Trainers.

Though NMF’s approach is more efficient, this inconsistency lowers code readability and makes it harder to refactor for multi-thread feature.
